//-------------------------------------------------------------------------------
// Name:        console.proto
// Purpose:     protobuf and grpc for remote control of navigation server
//
// Author:      Laurent Carré
//
// Created:     03/05/2022
// Copyright:   (c) Laurent Carré Sterwen Technology 2021-2023
// Licence:     Eclipse Public License 2.0
//-------------------------------------------------------------------------------

syntax = "proto3";

import "arguments.proto";

enum State {
     STOPPED=0;
     RUNNING = 1;
  }

message CouplerMsg {
  string  name=1;
  string coupler_class=2;
  State state=3;
  enum Device_state {
    NOT_READY=0;
    OPEN=1;
    CONNECTED=2;
    ACTIVE=3;
  }
  Device_state dev_state=4;
  string protocol=5;
  uint32 msg_in=6;
  uint32 msg_out=7;
  string status=8;
  uint32 error=9;
  float input_rate=10;  // messages / sec
  float output_rate=11;
}

message Connection {
  string remote_ip=1;
  uint32 remote_port=2;
  uint32 total_msg=3;
  float msg_rate=4;
}

message Server {
  string server_class=1;
  string name=2;
  string server_type=3;
  bool running=4;
  uint32 nb_connections=5;
  uint32 port=6;
  repeated Connection connections=7;
}

message NavigationServerMsg {
  uint32 id=1;
  string name=2;
  State state=3;
  string status=5;
  uint32 error=6;
  string version=7;
  string start_time=8;
  repeated Server servers=9;
}

message N2KDeviceMsg {
  uint32 address=1;
  uint32 manufacturer_code=2;
  string manufacturer_name=3;
  uint32 product_code=4;
  string product_name=5;
  string description=6;
}

message Request {
  uint32 id=1;
  string cmd=2;
  string target=3;
  ArgumentList kwargs=4;
}

message Response {
  uint32 id=1;
  string status=2;
  ArgumentList response_values=3;
}

service NavigationConsole {
  rpc GetCouplers(Request) returns (stream CouplerMsg) {}
  rpc GetCoupler(Request) returns (CouplerMsg) {}
  rpc CouplerCmd(Request) returns (Response) {}
  rpc ServerStatus(Request) returns (NavigationServerMsg) {}
  rpc ServerCmd(Request) returns (Response) {}
  rpc GetDevices(Request) returns (stream N2KDeviceMsg) {}
}

